### PR TITLE
Adjust gitconfig in case if git credentials file are going to be mounted into the workspace

### LINKED ISSUE #17251: Provide git credential storage mechanism
### Is your enhancement related to a problem? Please describe.

When using git with HTTPS protocol all the connection needs a username and password. To avoid prompting for credential at every connection git allows to configure a credential storage mechanism:

**cache with timeout**
```bash
$ git config --global credential.helper 'cache --timeout=3600'
``` 

**local file with cleartext username/password (never expires)**
```bash
git config --global credential.helper 'store --file ~/.git-store/credentials'
```

This is described [here](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage).

That can be done manually on che workspaces but requires a manual step (`git config` command) that needs to be repeated at every restart of the workspace.

### Describe the solution you'd like

We should document how to use the [secret injection mechanism](https://www.eclipse.org/che/docs/che-7/mounting-a-secret-as-a-file-or-an-environment-variable-into-a-workspace-container/) to inject a file with the username and password and set the git config `credential.helper` accordingly:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-credentials-secret
  annotations:
    che.eclipse.org/automount-workspace-secret: true
    che.eclipse.org/mount-path: /home/user/.git-store/
    che.eclipse.org/mount-as: file
  data:
     credentials: <base64 encoded file>
```

And in the git configuration:
```
[credential]
        helper = store --file ~/.git-store/credentials
```

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
Adjust gitconfig in case if git credentials file are going to be mounted into the workspace

### What issues does this PR fix or reference?
Fixes https://github.com/eclipse/che/issues/17251

#### Release Notes
n/a

#### Docs PR
https://github.com/eclipse/che-docs/pull/1421

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java:2]: mine is making `2018` too.. lol

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An instance of {@link FileSecretApplier} that is trying to adjust the content of git-config that
+ * was added by {@link GitConfigProvisioner}. The adjustment is adding configuration of git
+ * credentials store, which is pointing to the file that is going to be mount to the container from
+ * the secret.
+ */
+public class GitCredentialStorageFileSecretApplier extends FileSecretApplier {
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java:41]: `@Singleton`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            .append('\n')
+            .append("[credential]")
+            .append('\n')
+            .append('\t')
+            .append("helper = store --file ")
+            .append(gitSecretFilePath.toString())
+            .append('\n');
+        HashMap<String, String> newGitConfigMapData = new HashMap<>(gitConfigMapData);
+        newGitConfigMapData.put(GitConfigProvisioner.GIT_CONFIG, gitConfigBuilder.toString());
+        LOG.info(gitConfigBuilder.toString());
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java]: can be debug imo

**Review by mshaposhnik — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    if (gitConfigMap != null) {
+      Map<String, String> gitConfigMapData = gitConfigMap.getData();
+      String gitConfig = gitConfigMapData.get(GitConfigProvisioner.GIT_CONFIG);
+      if (gitConfig != null) {
+        if (gitConfig.contains("helper = store --file ") && gitConfig.contains("[credential]")) {
+          throw new InfrastructureException(
+              "Multiple git credentials secrets found. Please remove duplication.");
+        }
+
+        StringBuilder gitConfigBuilder = new StringBuilder(gitConfig);
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java]: Simple string concatenation would be faster and IMO more readable here.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    Path gitSecretFilePath = Paths.get(secretMountPath, keys.iterator().next());
+    ConfigMap gitConfigMap =
+        env.getConfigMaps()
+            .get(
+                runtimeIdentity.getWorkspaceId() + GitConfigProvisioner.GIT_CONFIG_MAP_NAME_SUFFIX);
+    if (gitConfigMap != null) {
+      Map<String, String> gitConfigMapData = gitConfigMap.getData();
+      String gitConfig = gitConfigMapData.get(GitConfigProvisioner.GIT_CONFIG);
+      if (gitConfig != null) {
+        if (gitConfig.contains("helper = store --file ") && gitConfig.contains("[credential]")) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java]: is the whitespace at the end of `"helper = store --file "` intentional?

**Review by sparkoo — APPROVED** [review]
lgtm. 2 minor comments

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java (hunk 5 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/GitCredentialStorageFileSecretApplier.java]: no, just copy-paste. Let's remove it.

**Pull request merged**